src/enhanced/utils/queue/index.ts

```diff
@@ -0,0 +1,48 @@
+import { SQS } from '@aws-sdk/client-sqs';
+import config from 'config';
+import { JobsQueuePayload } from '../../types/job-types';
+import { isJobQueuePayload } from '../../types/utils/type-guards';
+
+/**
+ * This function listens to the job queue until a message is received.
+ * @returns {Promise<JobsQueuePayload>} the promise for the payload object after a message has been received
+ */
+export async function listenToJobQueue(): Promise<JobsQueuePayload> {
+  const region = config.get<string>('aws_region');
+  const queueUrl = config.get<string>('jobsQueueUrl');
+
+  const client = new SQS({ region });
+
+  // We want to loop indefinitely so that we continue to poll the queue.
+  while (true) {
+    const receiveMessage = { QueueUrl: queueUrl, MaxNumberOfMessages: 1, WaitTimeSeconds: 2 };
+
+    const res = await client.receiveMessage(receiveMessage);
+
+    if (res.Messages && res.Messages.length > 0) {
+      const message = res.Messages[0];
+
+      if (!message.Body) {
+        console.error(`ERROR! Received message from queue without body. Message ID is: ${message.MessageId}`);
+        continue;
+      }
+
+      const payload = JSON.parse(message.Body);
+
+      // Use type guard here to validate payload we have received from the queue.
+      // This ensures that the `payload` object will be of type `JobQueuePayload` after the if statement.
+      if (!isJobQueuePayload(payload)) {
+        console.error(
+          `ERROR! Invalid payload data received from message ID: ${
+            message.MessageId
+          }. Payload received: ${JSON.stringify(payload)}`
+        );
+        continue;
+      }
+
+      // Great! we received a proper message from the queue. Return this object as we will no longer
+      // want to poll for more messages.
+      return payload;
```

Comment:
QQ: i'm not sure if this would be the right place, but would we want to delete message after it has been received, or does this happen at the end of the task? did not see anywhere where we are deleting queue messages, perhaps the message is processed (and marked for deletion?)

according to SQS docs on `visibility`, `If the consumer fails to process and delete the message before the visibility timeout expires, the message becomes visible to other consumers.`

Reply:
Ah, good catch. I assumed the message would be removed from the queue if it has been received, but that isn't correct. I'll update my code to handle deletion of a message tomorrow👍

Reply:
I think it's how SQS and Lambda interact. When a message is read from an SQS queue, and the lambda is triggered, I believe that handles the message deletion on its own. We are using the client here which requires us to do a bit more work I believe to achieve the same behavior.